content/posts/2022-04-bundling-nodejs-for-aws-lambda.md

```diff
@@ -0,0 +1,495 @@
+    ),
+  };
+
+  const moduleDeps = [];
+  for (const { importPath, resolvedImportPath } of imports) {
+    if (importPath.startsWith('./') || importPath.startsWith('../')) {
+      // relative import of a single project-local file
+      const localPath = path.relative(projectRoot, resolvedImportPath);
+      archiveItems[localPath] = new pulumi.asset.FileAsset(resolvedImportPath);
+    } else {
+      // module import, may be resolved to a package cache - save it for later
+      moduleDeps.push(importPath);
+    }
+  }
+
+  // TODO for each dependency in `moduleDeps`:
+  // find all its files, including transitive dependencies, and add them to archiveItems.
+
+  return new pulumi.asset.AssetArchive(archiveItems);
+}
+```
+
+Collecting files of all transitive dependencies is relatively easy thanks to the
+pnpm module layout because every module has its dependencies linked next to it.
+It's about ~40 lines of code; the essential part is to call `fs.realpath` to
+resolve symlinks into their actual location in the pnpm cache.
+
+This solution worked pretty well for a while. Until I started to add more
+dependencies to my lambda. Looking at you,
+[Postgraphile](https://www.graphile.org)! At that point, I exceeded the maximum
+allowed size for Lambda code, and it was again time to look for yet another
+solution.
+
+## esbuild
+
+At this point, I gave up. I decided to follow what others are doing and adopt
+[esbuild](https://esbuild.github.io). I am not happy that I am running a
+different code from what I see in my project, but I guess that's the price I
+have to pay for using AWS Lambda as the deployment target.
+
+At a high level, esbuild makes bundling super easy: you give it a path your
+JavaScript (or even TypeScript!) handler file and esbuild gives you back a
+single JavaScript file containing both your code and all dependencies. You can
+even use `pulumi.asset.StringAsset` to keep everything in memory.
+
+Of course, things are more complicated in practice.
+
+1. Top-level await requires a tweak in esbuild config.
+2. `StringAsset` comes with a performance penalty when the content is large.
+3. I want source map support for Error stack traces.
+4. The client `pg` has an optional dependency `pg-native`, which I am not
+   installing. esbuild fails when this dependency is not found.
+5. The code produced by esbuild for Node.js ESM target does not work in several
+   edge cases.
+6. Code resolving paths using `__dirname` and `import.meta.url` does not work
+   when bundled.
+
+### Top-level await
+
+I am running my code on the latest Node.js version 14.x provided by AWS Lambda;
+this version supports top-level await. Support for top-level await is not
+available in early 14.x versions; therefore, we must tell esbuild to target
+Node.js 14.8 or later.
+
+```js
+/** @type {esbuild.BuildOptions} */
+const esbuildConfig = {
+  // (...)
+  platform: 'node',
+  target: 'node14.8',
+  format: 'esm',
+};
+```
+
```

Comment:
Note: by the time this blog post is published, AWS Lambda added support for Node.js 16. I am using `target: 'node16'` now.